### ISSUE TITLE
pam_faillock default of deny=3 is too stringent

### ISSUE THREAD

**Issue by AnilRedshift** [linux-pam/linux-pam#420]
I created #419 to change the default to, say 20.

Three failed passwords in the window is so incredibly easy to do, especially with a complicated password, or with multiple keyboard layouts.

Here are some downstream examples of people complaining about the default: https://bugs.archlinux.org/task/67644

**Comment by ldv-alt** [comment]
Looks like the main point of complains in the Arch Linux bug report is not that `3` is too stringent default for `deny=` option, but that the default action of denying access for a period of time is too stringent.
Unfortunately, the defaults cannot make everybody happy, but, fortunately, pam_faillock is configurable.

**Comment by AnilRedshift** [comment]
I think it can be both - the threshold for lockout is too low and the action taken is not proportionate.
From a user's perspective, I think it is quite often possible to mistype a password 3 times within the window.
As an personal anecdote, if I have my keyboard layout switched to the wrong layout, then my first password will be incorrect. Usually I think it's a typo so I'll try typing it in again.

Then when it doesn't work, I would realize it's in the wrong layout and switch. Now I only have one attempt to log in correctly.

I can think of many other cases when users would have difficulties - such as using a touch keyboard, having caps-lock pressed etc.

For some back-of-the-napkin math: If 5% of keystrokes have a chance of being wrong AND you not noticing, then you end up with a 10% chance of being locked out: `(1 - (0.95 ^ 12)) ^3`

a 98% accuracy would give you a lockout rate of 1%, or several times a year.

Compare that to changing the default to 20 attempts. You would end up with being locked out accidentally ~ 0 times a year ` `(1 `- (0.95 ^ 12)) ^20`

From a security perspective, even for something like a 4 digit pin, changing the lockout from 3 -> 20 would mean that you have 
3 / 10,000 chance of guessing the right number (e.g. 0.03% chance). Changing that to a threshold of 20 would mean a probability of 0.2%. And that's with a 4 digit pin (again back of the napkin, ignoring things like unequal password distribution)

TL;DR: I think there's a strong case to be made for raising the threshold to something larger.

**Comment by notlesh** [comment]
Learning a new password, new keyboard layout, switching keyboards, etc. will all lead to a high probability of triggering 3 failures within a short window.

**Comment by t8m** [comment]
Why the Arch Linux distro can ship a PAM configuration with pam_faillock set up by default but they cannot ship a faillock.conf file that would change the default to 20?
IMO there is no point in changing the default when it can be easily configured to a different value by the distro or by the sysadmin?

**Comment by AnilRedshift** [comment]
The reason for changing the default is that it's causing pain for users. I've made an argument why I think that 3 is a bad, real-life default. I think the best discourse here would be to figure out the following question, and from there we can find the right place for a solution.

Question: Is the default of 3 a "good" default for pam_faillock?
Possible Answers:
* Yes - and here are some reasons why
* Sometimes - and here are some reasons why - and if you don't like it then change the `.conf`
* No - and here are some reasons why.

In my (previous post)[https://github.com/linux-pam/linux-pam/issues/420#issuecomment-1003454342], I gave my answers (subjective & objective) for the "No" bucket.

I'd love to hear from you, or other folks a rationale for a different answer. We can come to an agreement on the nature of the problem and then move to the correct solution from there.

Thanks!

**Comment by t8m** [comment]
My answer: Sometimes and if you don't like it then change the .conf

The reason(s): The built-in default is just this - a built-in default. It does not mean that it needs to satisfy all use-cases. One of the use cases - the one that it was initially written for is compliance with Common Criteria certification rules. There the defaults are adequate as the security requirements in CC are at quite paranoid level. And yes, these defaults do not make much sense for regular workstation installations but there most probably the pam_faillock does not make sense to be placed in the default PAM stack configuration at all.

**Comment by hollunder** [comment]
For a year now I wondered why my sudo pw was not accepted even though it was correct. Just yesterday it happened to me on the initial login and was only fixed after power cycling. I thought my system is seriously broken.
This is an absolutely stupid change of defaults.

**Comment by notlesh** [comment]
> it was initially written for is compliance with Common Criteria certification rules.

That's a compelling reason, but it serves a very small minority of users. Is there perhaps a way to make this work for CC certification but not be such a nuisance?

> For a year now I wondered why my sudo pw was not accepted even though it was correct. Just yesterday it happened to me on the initial login and was only fixed after power cycling. I thought my system is seriously broken. This is an absolutely stupid change of defaults.

I actually thought something was horribly broken for a while as well.

One of the problems is that, when using SSH or sudo, you don't actually get any meaningful error message. The UX is indistinguishable from just typing the wrong password (or any number of other symptoms, like logging into the wrong machine, using the wrong kb layout, etc. etc. etc.). It wasn't until I tried logging in via getty that I saw a useful error message.

I suspect the answer is "that's downstream's problem," but is there a way we could improve the error message when this happens?

**Comment by t8m** [comment]
There was no change in defaults for pam_faillock! That Arch Linux decided to put pam_faillock as a module that is present by default in the configuration and not change the default either by patching the code or providing a default config file with some settings that would be more reasonable for general use cases, is their fault! I am going to close this issue.

**Comment by traverseda** [comment]
Generally the arch policy is to try and modify upstream as little as possible. I think that's a very *good* policy, presuming that the upstream developers are committed to setting up sane defaults.

Now I'm not sure that pam is intended to ship with faillock enabled by default. Is it? Do you (the maintainers of pam) think that faillock should be enabled by default? That it should be enabled with those defaults by default?

Because if you don't having you say that will actually go a long way towards getting the arch linux team to not include it by default, they do respect upstream and how upstream intends the packages to be used. If you could clarify when you think it's appropriate to use faillock, and what some appropriate defaults would be for it, I think a lot of users would appreciate it. 

Personally this problem has been a problem for me off and on for a while, but I didn't care enough to track it down to this issue until I switched to a new laptop a few days ago, one with a subtly different keyboard layout. Obviously I've fixed the issue on my laptop but I think it would save a lot of people a lot of frustration if we could get either some better defaults or possibly a more descriptive error message when it happens?

**Comment by polarathene** [comment]
**Summary of verbose response:**

- As **defaults are subject to change**... `deny = 3` as a default (_justified to satisfy paranoid levels of compliance_) seems odd? Any security audit that relies on that is **likely to mandate explicit configuration** anyway?
- Unless the majority of deployments where `pam_faillock` is intended to be used must comply with CC rules, and the default is somehow required to be certified for such - adjusting the default to suit the majority of use-cases where `pam_faillock` is useful would be appreciated.
- There are numerous scenarios where `deny = 3` is problematic - especially when the user is not a security professional and trusts the maintainers to know what is most appropriate.
- The docs for the module do not clarify that the configuration defaults (and example config shown) are tailored to paranoid levels, nor that the module is discouraged for systems running a desktop environment.

If adjusting the default value for `deny` is out of the question... **Could you at least** be open to improving the modules docs?:
- At least with a statement about when the module is suitable (or not).
- What the defaults and example config are intended for (_alternatively an example that is suitable for non-paranoid contexts, such as desktop/workstation systems_).

---

<details>
<summary>Verbose response</summary>

> The reason(s): The built-in default is just this - a built-in default. It does not mean that it needs to satisfy all use-cases.

:+1: That is very true. But I think we can agree that defaults ideally are sane/good for the general use-cases that it supports? (_and can usually be revised to a more appropriate default when it makes sense to?_)

https://github.com/linux-pam/linux-pam/blob/510e825ef130a843663115ec510b1237ea4708f4/modules/pam_faillock/faillock.conf#L29-L32

> One of the use cases - the one that it was initially written for is compliance with Common Criteria certification rules.
> There the defaults are adequate as the security requirements in CC are at quite paranoid level.
> And yes, these defaults do not make much sense for regular workstation installations

Does that certification require a default to meet compliance? Or just that it is configurable to meet compliance?

I would assume that in security contexts, it's important that the admin knows how to configure a system properly to meet any compliance requirements. Would they trust defaults implicitly? Would they be opposed to `pam_faillock` having a higher `deny` default?

Contrast that to other use-cases of `pam_faillock`, who benefits the most from practical defaults where `pam_faillock` can be useful?

In the case of Arch, they try to minimize adding any custom configuration by trusting the defaults from upstream where possible. There are no alternative editions catering specifcally for server or desktop users, and like the policy to respect the default configs, this is something users of Arch generally prefer (_assuming upstream has sane defaults_).

---

I am curious if the majority of deployments with `pam_faillock` would be actually need to comply with Common Criteria rules, and expect such compliance be provided by default configs with all software, rather than expecting to be configured for paranoid levels explicitly when necessary?

---

> That Arch Linux decided to put `pam_faillock` as a module that is present by default in the configuration and not change the default either by patching the code or providing a default config file with some settings that would be more reasonable for general use cases, is their fault!

I think in all fairness, Arch maintainers trusted your defaults to be sane (_tailored for the majority of use-cases `pam_faillock` was intended for_), not for paranoid level of compliance as you mentioned.

Your docs for the module also provide a similar example with very low `deny`:

https://github.com/linux-pam/linux-pam/blob/510e825ef130a843663115ec510b1237ea4708f4/modules/pam_faillock/faillock.conf.5.xml#L201-L223

No where in these docs have you implied that these defaults are aimed at meeting "Common Criteria rules" for compliance at paranoid levels.

Would it not be better to provide an example config (_or alternatively an explicit statement against using the module_) regarding what maintainers of the module consider suitable for workstation use?

Hopefully that's not too much to ask (better communication), especially when it comes to security focused projects, often users (_and this perhaps includes maintainers of Arch too_) will trust the experts upstream to know what is best unless advising users otherwise.

Whereas security professionals likely know better and know how to configure software for their security requirements, especially when they have compliance requirements to meet. It seems to be the wrong demographic to bias defaults to, unless absolutely required to meet that compliance (_seems unlikely since you advise downstream to adjust defaults, and a deployment may inherit those instead, requiring configuration anyway, at least in awareness and to check/verify the default_).

Additionally, **as defaults can be subject to change**...for compliance reasons I can only imagine that **an audit would expect `deny` to be configured explicitly?** That seems like a fairly compelling reason that `deny = 3` default for compliance seems like an odd choice?

---

For what it's worth, [Arch does document this issue for users](https://wiki.archlinux.org/title/Security#Lock_out_user_after_three_failed_login_attempts), getting a change for Arch to configure `pam_faillock` (or disable it) AFAIK relies on you to make it clear via documentation, otherwise like you've responded, they defer us to their docs due to their own policy to respect upstream (_with the intent that users on Arch get defaults from upstream, usually to avoid surprise changes when troubleshooting like other distros impose_).

I have no issue adjusting configuration myself, but considering the number of users that have negative experiences (regardless of Arch specifically), I wanted to add another voice of reason to consider the feedback as something actionable by `linux-pam`.

---

> And yes, these defaults do not make much sense for regular workstation installations but there most probably the `pam_faillock` does not make sense to be placed in the default PAM stack configuration at all.

It would be great if you can add that to the docs for the module :+1: 

Is it your opinion that `pam_faillock` is not worthwhile for most Desktop users security under any configuration? Is it not intended to protect a desktop user in any meaningful way if `deny` was adjusted to something more tolerable?

What about servers where security is usually more important, is `deny = 3` acceptable? Or would you suggest not using `pam_faillock` there either, or bumping it up to what value?

</details>

**Comment by t8m** [comment]
It is very much questionable whether pam_faillock should be enabled at all for average user configuration. Thus no, we are not going to change the default configuration.

**Comment by polarathene** [comment]
> It is very much questionable whether pam_faillock should be enabled at all for average user configuration.

- Is that advised anywhere in the docs? Or are users of `linux-pam` meant to implicitly know what is discouraged for the average user?
- Is there no sane reason to use `pam_faillock` for the average user under any configuration?
- Is it only worthwhile to a different audience, and would that audience ever have reason to relax the config from the present defaults? (_which has been tailored for paranoid levels_)

---

Perhaps without officially documenting that `pam_faillock` is discouraged for the average user (_regardless of configuration for it_), there is some confusion where it might seem relevant to an average user?

My Android phone, and likely many other such devices for average users by default provides a limited amount of unlock attempts before it locks access for a duration.

If that's not serving effectively the same purpose on a non-desktop device, it's easy to see how one might think `pam_faillock` is the desktop equivalent, and with a more relaxed config would not cause as much frustration to an average desktop user that may want/appreciate such a feature?

**Comment by mohemiv** [comment]
When a user starts a command with sudo and then realizes that it's no longer necessary to run it, they usually press Ctrl+C multiple times.

After that, the user is locked out of the system and cannot do anything for 10 minutes.

If a graphical session is locked, the user will need to perform a hard reboot of the system to regain access if time is critical. They may also lose some of their data.

**Comment by Iiridayn** [comment]
I came here to list a few research papers such as "Ten strikes and you’re out: Increasing the number of login attempts can improve password usability" by Brostoff and Sasse, but it looks like the NIST is already aware of this research and have updated their criteria to match as mentioned in #641.

Thank you for this module, and for mentioning the Common Criteria certification rules as an inspiration. I have been searching them, and though this module appears to meet the FIA_AFL family, I have not found any value specified in the Common Criteria certification rules; just that there should _be_ a value (and that the account or terminal has to be locked, etc). Would you please help me see where they specify that 3 is the required value? I've found FIA_AFL in https://www.commoncriteriaportal.org/files/ccfiles/CC2022PART2R1.pdf (page 88) and https://www.commoncriteriaportal.org/files/ccfiles/cc_2022.xml so far; is there another document with specific numbers?

If there is not, would you please consider bending to security best practices and establishing a default value of 10, as recommended by NIST, Brostoff and Sasse, and other security researchers?

If you help me see that they do mandate exactly 3, I will work with them to update their value to match security best practices, and then return once that is done.

**Comment by Iiridayn** [comment]
- https://www.commoncriteriaportal.org/files/ppfiles/authpp[1].pdf (page 19) specifies a value, though that is "1" instead of 3.
- https://www.commoncriteriaportal.org/files/ppfiles/PP9906.pdf (page 23) specifies a range from 1-5, though notes the ST author has to pick something.

I found https://www.tksg.global/pluginfile.php/55702/mod_resource/content/0/AIS_41_BSI_PP_ST_Guide_pdf.pdf helpful. It appears there is the CC spec, which gives menus of options, then those are put together into Protection Profiles (PPs) which are restricted menus of what customers want leaving much undefined. Developers in turn produce Security Targets (STs) which say what they did, and must conform to the PP. The software can then be tested to conform to the ST.

The Protection Profiles I've found typically do not appear to prefer to list specific values for FIA_AFL, likely delegating the choice to the developers. Regardless, this specification process quite clearly implies, as suggested by another user, that there is no "magic number" which would satisfy every client - therefore, defaulting to the NIST requirement seems quite reasonable. "Ten strikes and you’re out: Increasing the number of login attempts can improve password usability" finds that 47% of users would be inconvenienced by a 3-strikes rule, while only 7% would be inconvenienced by a 10-strike rule. Explains why NIST changed their guidelines.

Quantifying security differences is challenging, as user passwords tend to conform to Zipf's law - the attacker has a huge early advantage which tapers off. Randomly generated passwords intended to meet a specified security level would only need to add 2.8 bits, which is equivalent to adding symbols to an 8 character alphanumeric password, or a single digit to a 8-digit PIN (now 9). 2.8 bits isn't nothing, but there are a number of ways to inject more security - see Passwords and Cryptwords: The Final Limits on Lengths for a discussion.

Effectively, if we assume we give an attacker a year of DoS lockouts on an account, with a 10-minute lockout as default, they can try around 525949 instead of 157785 passwords per year. An Administrator’s Guide to Internet Password Research for the suggestion that 6 digit PINs are sufficient for authentication passwords; if randomly generated these the odds of a successful guess during that year rise from around 16% to around 53%. For longer passwords the proportionate increase is still 10/3=3.3, but the total odds remain low. If the account has any value, being effectively constantly locked out for a year should be detectable.

"So long, and no thanks for the externalities: the rational rejection of security advice by users" suggests user time is worth "$2.6 billion an hour". I believe most users and cost-sensitive IT managers would prefer 10 strikes. Even security sensitive IT managers should recognize that conforming to best practices is a good move.

**Comment by ioquatix** [comment]
I'm all for a reasonable low limit on password attempts, e.g. 3-5 could be reasonable.

However, I'd like to clarify that I don't think the glitch people are experiencing is due to failed password entries - in my experience, simply showing the password box on the screensaver is sufficient to consume an attempt. If someone bumps the keyboard several times, causing the password entry box simply to be shown, that can count against the limit. If the software is buggy, and the password box is shown several times with no entry, this can cause the system to become locked, which is extremely inconvenient.

Therefore, I think we should be clear on what constitutes a failed attempt: at least 1 character should be submitted in order to count against the limit.

**Comment by Iiridayn** [comment]
@ioquatix I'm happy to move my comments to #641, though that one was closed and pointed to this one. I feel having an appropriate security default is a different issue than bugs related to login failures.
